UI/CUIControls.cpp

```diff
@@ -1930,61 +1931,65 @@ void ProductionInfoPanel::DoLayout() {
+MultiTurnProgressBar::MultiTurnProgressBar(int num_segments,
+                                           float percent_completed,
+                                           float percent_predicted,
+                                           const GG::Clr& bar_color,
+                                           const GG::Clr& bg_color,
                                            const GG::Clr& outline_color) :
     Control(GG::X0, GG::Y0, GG::X1, GG::Y1, GG::NO_WND_FLAGS),
-    m_total_turns(std::max(1, total_turns)),
-    m_turns_completed(std::max(0.0, std::min<double>(turns_completed, m_total_turns))),
-    m_total_cost(total_cost),
-    m_turn_spending(turn_spending),
-    m_bar_color(bar_color),
-    m_background(background),
-    m_outline_color(outline_color)
-{}
-
-void MultiTurnProgressBar::Render() {
-    GG::Pt ul = UpperLeft(), lr = LowerRight();
-    bool segmented = true;
-    if (Width() / m_total_turns < 3)
-        segmented = false;
-
-    // draw background over whole area
-    FlatRectangle(ul, lr, m_background, m_outline_color, 1);
-
-    // draw predicted bar of next turn
-    GG::X completed_bar_width(std::max(1.0, Value(Width() * m_turns_completed / m_total_turns)));
-    GG::X predicted_bar_width(Value(Width() * (m_turn_spending / m_total_cost)));
-
-    if (predicted_bar_width > 3) {
-        GG::Pt predicted_bar_ul(ul.x + completed_bar_width - 1, ul.y);
-        GG::Pt predicted_bar_lr;
-        if ((completed_bar_width + predicted_bar_width) > Width()) // cut if beyond right border
-            predicted_bar_lr = GG::Pt(ul.x + Width(), lr.y);
-        else
-            predicted_bar_lr = GG::Pt(ul.x + completed_bar_width + predicted_bar_width, lr.y);
-        FlatRectangle(predicted_bar_ul, predicted_bar_lr, GG::LightColor(m_bar_color), m_outline_color, 1);
-    }
-
-    // draw completed portion bar
-    if (completed_bar_width > 3) {
-        GG::Pt bar_lr(ul.x + completed_bar_width, lr.y);
-        FlatRectangle(ul, bar_lr, m_bar_color, m_outline_color, 1);
+    m_num_segments(num_segments),
+    m_perc_completed(percent_completed),
+    m_perc_predicted(percent_predicted),
+    m_clr_bar(bar_color),
+    m_clr_bg(bg_color),
+    m_clr_outline(outline_color)
+{
+    // validate percentage values
+    if (m_perc_completed < 0.0f || m_perc_predicted < 0.0f ||
+        (m_perc_completed + m_perc_predicted) > 1.0f)
+    {
+        WarnLogger() << "Values not within percent range, clamping: "
+                     << m_perc_completed << ", " << m_perc_predicted;
+        boost::algorithm::clamp(m_perc_completed, 0.0f, 1.0f);
+        boost::algorithm::clamp(m_perc_predicted, 0.0f, 1.0f);
     }
+}
 
-    // draw segment separators
-    if (segmented) {
+void MultiTurnProgressBar::Render() {
+    GG::Pt ul(UpperLeft());
+    GG::Pt lr(LowerRight());
+    GG::Y bottom(lr.y);
+    // draw background
+    FlatRectangle(ul, lr, m_clr_bg, m_clr_outline, 1);
```

Comment:
would be preferable to render with vertex arrays here rather than a series of calls to FlatRectangle which I think will end up in here https://github.com/freeorion/freeorion/blob/master/GG/src/DrawUtil.cpp#L54 doing separate calls to set the rendering state each time.